File: content/snippets/go/design_patterns_prototype.go
package main

import "fmt"

type Prototype interface {
	Clone() Prototype
	Describe()
}

type Report struct {
	Title string
	Tags  []string
}

func (r Report) Clone() Prototype {
	tags := append([]string(nil), r.Tags...)
	return Report{Title: r.Title, Tags: tags}
}

func (r Report) Describe() {
	fmt.Printf("%s %v\n", r.Title, r.Tags)
}

func main() {
	template := Report{Title: "weekly", Tags: []string{"ops", "summary"}}
	copy := template.Clone()
	template.Describe()
	copy.Describe()
}

File: content/snippets/scala3/design_patterns_prototype.scala
// Prototype
trait Prototype:
    def clone(): Prototype

case class Document(title: String, content: String) extends Prototype:
    def clone(): Document = this.copy()

@main def main(): Unit =
    val original = Document("Report", "Initial content")
    val copy1 = original.clone()
    val copy2 = original.clone()
    
    println(original)
    println(copy1)
    println(copy2)
